### PR TITLE
feat: nuxt module

### PR DESCRIPTION
**Pull request by mattmaribojoc**
# Related issue

https://vsf.atlassian.net/browse/SFUI2-911

# Scope of work

Adds a Nuxt Module that:
- enables autoimports of components and composables
- Sets up default Tailwind configuration for Storefront UI to work (preset and content)
  - The content path is overridable via Nuxt config
- Installs `@nuxtjs/tailwindcss` module if it isn't found already   

# Screenshots of visual changes

# Notes
was getting lint errors on the rule `vue/no-dupe-keys`, tracked it back to [this issue ](https://github.com/vuejs/eslint-plugin-vue/issues/2096). But added an ignore for this PR.

# Checklist

- [x] Self code-reviewed
- [x] Changes documented
- [x] Semantic HTML
- [x] SSR-friendly
- [x] Caching friendly
- [x] a11y for WCAG 2.0 AA
- [x] examples created
- [x] blocks created
- [x] cypress tests created

### CODE REVIEW

File: apps/preview/nuxt/tailwind.config.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { tailwindConfig } from '@storefront-ui/vue/tailwind-config';
 
 export default <Config>{
   presets: [tailwindConfig],
-  content: ['./**/*.vue', '../../../packages/sfui/frameworks/vue/**/*.vue'],
+  content: ['./**/*.vue', './..../../../packages/sfui/frameworks/vue/**/*.vue'],
```
Comment (FRSgit) [inline apps/preview/nuxt/tailwind.config.ts:7]: typo? 😄

File: packages/sfui/frameworks/nuxt/package.json [hunk 2]
```diff
@@ -0,0 +1,45 @@
+{
+  "name": "@storefront-ui/nuxt",
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/package.json:2]: Just wanted to raise one question - shouldn't we call it @storefront-ui/nuxt-module? `nuxt` seems a bit too generic - but maybe it's only me

File: packages/sfui/frameworks/nuxt/package.json [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    "dev:prepare": "nuxt-module-build --stub && nuxi prepare playground",
+    "release": "npm run lint && npm run test && npm run prepack && changelogen --release && npm publish && git push --follow-tags"
+  },
+  "dependencies": {
+    "@nuxt/kit": "^3.3.2",
+    "@nuxtjs/tailwindcss": "^6.6.4",
+    "@storefront-ui/vue": "workspace:*"
+  },
+  "devDependencies": {
+    "@nuxt/eslint-config": "^0.1.1",
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/package.json:40]: I'd prefer to use the same eslint config as in other places of our repo, wdyt?

File: packages/sfui/frameworks/nuxt/package.json [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  "types": "./dist/types.d.ts",
+  "files": [
+    "dist"
+  ],
+  "scripts": {
+    "build:nuxt-module": "nuxt-module-build",
+    "dev": "nuxi dev playground",
+    "dev:build": "nuxi build playground",
+    "dev:prepare": "nuxt-module-build --stub && nuxi prepare playground",
+    "release": "npm run lint && npm run test && npm run prepack && changelogen --release && npm publish && git push --follow-tags"
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/package.json:32]: Is `changelogen` supporting monorepos? I think we should remove this one for now and come up with proper changelog generator for all of the projects in the repo

File: packages/sfui/frameworks/nuxt/src/module.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+    // @ts-ignore
+    nuxt.options.tailwindcss = {
+      // @ts-ignore
+      ...nuxt.options.tailwindcss,
+      config: {
+        presets: [tailwindConfig],
+        content: [contentPath],
+        // @ts-ignore
+        ...nuxt.options.tailwindcss?.config,
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/src/module.ts:33]: Comment:
I think we should try to make it working with proper typings 😄 What's the issue here?

File: packages/sfui/frameworks/nuxt/.npmrc [hunk 6]
```diff
@@ -0,0 +1,2 @@
+shamefully-hoist=true
+strict-peer-dependencies=false
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/.npmrc:2]: Why is that needed?

File: packages/sfui/frameworks/nuxt/package.json (hunk 2 above)
Comment (mattmaribojoc) [reply packages/sfui/frameworks/nuxt/package.json:2]: The general approach that I was thinking of was the convention for these meta-framework specific themes would be `@storefront-ui/{meta-framework}/{framework}`.

So like:
- `@storefront-ui/nuxt`
- `@storefront-ui/astro/vue`
- `@storefront-ui/astro/react`

Specifically for this Nuxt module, I think having `nuxt` is the normal thing I've seen since this is what ends up in some's `nuxt.config.ts` modules.

So like 
```
export default defineNuxtConfig({
  modules: ['@storefront-ui/nuxt'],
})
```

File: packages/sfui/frameworks/nuxt/package.json (hunk 3 above)
Comment (mattmaribojoc) [reply packages/sfui/frameworks/nuxt/package.json]: leftover from the nuxt module starter :) - removing

File: packages/sfui/frameworks/nuxt/package.json (hunk 2 above)
Comment (FRSgit) [reply packages/sfui/frameworks/nuxt/package.json:2]: Okay, just wanted to make sure that there won't be anything wrong with naming package like this, thanks!

Regarding astro example:
[Official NPM package name validator](https://www.npmjs.com/package/validate-npm-package-name) says we cannot name the package `@storefront/astro/vue`.
image
So, if we would like to allow users to import things from `@storefront-ui/astro/react`and `@storefront-ui/astro/vue`  we would need to export these from a single package - `@storefront-ui/astro`. And I would not mix packages for different frameworks, it might lead to troubles with build and other things.

File: apps/preview/nuxt/tailwind.config.ts [hunk 7]
```diff
@@ -4,7 +4,7 @@ import { tailwindConfig } from '@storefront-ui/vue/tailwind-config';
 
 export default <Config>{
   presets: [tailwindConfig],
-  content: ['./**/*.vue', '../../../packages/sfui/frameworks/vue/**/*.vue'],
+  content: ['./**/*.vue', './../../../packages/sfui/frameworks/vue/**/*.vue'],
```
Comment (FRSgit) [inline apps/preview/nuxt/tailwind.config.ts:7]: Are you sure that this additional `./` is needed?
```suggestion
  content: ['./**/*.vue', '../../../packages/sfui/frameworks/vue/**/*.vue'],
```

File: packages/sfui/frameworks/nuxt/src/module.ts (hunk 5 above)
Comment (mattmaribojoc) [reply packages/sfui/frameworks/nuxt/src/module.ts:33]: `defineNuxtModule` is being imported from a hoisted dependency, but the `tailwindcss` key is dynamically generated based on the setup in the `playground`.

Some options would be `nuxt.options` through a validator like Zod to get the typechecks working, cast nuxt.options as any, or maybe look at not hoisting `@nuxt/kit`? 

Not sure would the best approach would be.

Comment (mattmaribojoc) [reply packages/sfui/frameworks/nuxt/src/module.ts]: 100a3271b45a8e6ba85992e423284fd05a8a7056

File: packages/sfui/frameworks/nuxt/playground/nuxt.config.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+          colors: {
+            primary: {
+              700: '#1e1e1e',
+            },
+          },
+        },
+      },
+    },
+  },
+} as any);
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/playground/nuxt.config.ts:26]: Why `as any` is needed?

File: packages/sfui/frameworks/nuxt/src/module.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    },
+  },
+  // Default configuration options of the Nuxt module
+  defaults: {
+    contentPath: './node_modules/@storefront-ui/vue/**/*.mjs',
+  },
+  async setup(options, nuxt) {
+    const { contentPath } = options;
+
+    const nuxtOptions = nuxt.options as unknown as NuxtConfig;
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/src/module.ts:25]: I cannot believe that this typing does not work out of the box 😄 but 🤷

**Comment by FRSgit** [comment]
Btw. is this still a `draft`?

**Comment by mattmaribojoc** [comment]
Still trying to resolve the TS issues that's why it's a draft. @FRSgit

File: packages/sfui/frameworks/nuxt/playground/nuxt.config.ts (hunk 8 above)
Comment (mattmaribojoc) [reply packages/sfui/frameworks/nuxt/playground/nuxt.config.ts]: Fix required bumping Nuxt version to `3.3` - updated preview app as well.

File: packages/sfui/frameworks/nuxt/src/module.ts (hunk 9 above)
Comment (mattmaribojoc) [reply packages/sfui/frameworks/nuxt/src/module.ts]: Tracked down the line causing issues to

```
declare module '@nuxt/schema' {
  interface NuxtConfig { ['tailwindcss']?: Partial<ModuleOptions> }
  interface NuxtOptions { ['tailwindcss']?: ModuleOptions }
}
```

`tailwindcss` uses NuxtOptions type without the Partial, because I think it's assuming this type would only be used by the tailwind module itself. 

I don't think nuxt/kit plays well with Typescript when one module is meant to influence the config of another.

Comment (FRSgit) [reply packages/sfui/frameworks/nuxt/src/module.ts]: Nice investigation! As long as it's not on our side - let's just stick with working code

File: packages/config/eslint/vue.js [hunk 10]
```diff
@@ -13,5 +13,6 @@ module.exports = {
     'vue/no-v-html': 0,
     'import/no-extraneous-dependencies': 0,
     'vue/multi-word-component-names': 0,
+    'vue/no-dupe-keys': 0
```
Comment (FRSgit) [inline packages/config/eslint/vue.js:16]:
```suggestion
    // needs to be reenabled after this gets merged: https://github.com/vuejs/eslint-plugin-vue/pull/2189 [evaluate after: 07-06-2023]
    'vue/no-dupe-keys': 0
```

File: packages/sfui/frameworks/nuxt/build.config.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import { defineBuildConfig } from 'unbuild';
+
+export default defineBuildConfig({
+  // If entries is not provided, will be automatically inferred from package.json
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/build.config.ts:4]:
```suggestion
```

File: packages/sfui/frameworks/nuxt/build.config.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+  entries: [
+    // default
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/build.config.ts:6]:
```suggestion
```

File: packages/sfui/frameworks/nuxt/package.json [hunk 13]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "@storefront-ui/nuxt",
+  "version": "1.0.0",
+  "homepage": "https://docs.storefrontui.io/v2/",
+  "installConfig": {
+    "hoistingLimits": "workspaces"
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/package.json:6]: I guess that's needed? :(

File: packages/sfui/frameworks/nuxt/playground/package.json [hunk 14]
```diff
@@ -0,0 +1,4 @@
+{
+  "private": true,
+  "name": "@storefront-ui/nuxt-playground"
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/playground/package.json:3]: What it this playground? What do we use it for? Only local development? Or maybe somewhere in the documentation?

File: packages/sfui/frameworks/nuxt/src/module.ts [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    nuxtOptions.tailwindcss = {
+      ...nuxtOptions.tailwindcss,
+      config: {
+        presets: [tailwindConfig],
+        content: [contentPath ?? ''],
+        ...nuxtOptions.tailwindcss?.config,
+        // if content is already defined, we need to merge it with the new one
+        ...(Array.isArray(nuxtOptions.tailwindcss?.config?.content) && contentPath
+          ? { content: [...nuxtOptions.tailwindcss?.config?.content!, contentPath] }
+          : {}),
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/src/module.ts:36]: I'd propose to rewrite this part in regular ifs - imo it's a bit more readable 

```suggestion
   const content = [];
   
   if (Array.isArray(nuxtOptions.tailwindcss?.config?.content?.files)) {
     content.push(...nuxtOptions.tailwindcss?.config?.content?.files);
   }
   if (Array.isArray(nuxtOptions.tailwindcss?.config?.content)) {
     content.push(...nuxtOptions.tailwindcss?.config?.content);
   }
   if (contentPath) content.push(contentPath);

    nuxtOptions.tailwindcss = {
      ...nuxtOptions.tailwindcss,
      config: {
        presets: [tailwindConfig],
        content,
```

File: packages/sfui/frameworks/nuxt/src/module.ts [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    if (!nuxt.options.modules.includes('@nuxtjs/tailwindcss')) {
+      await installModule('@nuxtjs/tailwindcss');
+    }
+
+    const components: string[] = [];
+    const composables: string[] = [];
+
+    Object.keys(storefrontUi).forEach((key) => {
+      // @ts-expect-error - checking package export to see if it's a component
+      if (key.startsWith('Sf') && (storefrontUi[key].__name || storefrontUi[key].name)) {
```
Comment (FRSgit) [inline packages/sfui/frameworks/nuxt/src/module.ts:49]: I understand this part, but it's pretty tricky. Maybe comment should be more straightforward like - "if object has a `name` or `__name` fields we consider it a Vue component instance"

File: packages/sfui/frameworks/nuxt/package.json (hunk 13 above)
Comment (mattmaribojoc) [reply packages/sfui/frameworks/nuxt/package.json:6]: yeah - without this, there were even more problems with `eslint` and some other dependencies coming from `nuxt/kit`

File: packages/sfui/frameworks/nuxt/playground/package.json (hunk 14 above)
Comment (mattmaribojoc) [reply packages/sfui/frameworks/nuxt/playground/package.json:3]: local development, it's used to test the module and also generate the types used for development. it's part of the nuxt module boilerplate.

File: packages/config/eslint/vue.js (hunk 10 above)
Comment (mattmaribojoc) [reply packages/config/eslint/vue.js:17]: 100a3271b45a8e6ba85992e423284fd05a8a7056

File: packages/sfui/frameworks/nuxt/build.config.ts (hunk 12 above)
Comment (mattmaribojoc) [reply packages/sfui/frameworks/nuxt/build.config.ts]: 100a3271b45a8e6ba85992e423284fd05a8a7056

File: packages/sfui/frameworks/nuxt/src/module.ts (hunk 15 above)
Comment (mattmaribojoc) [reply packages/sfui/frameworks/nuxt/src/module.ts:46]: 100a3271b45a8e6ba85992e423284fd05a8a7056

**Review by FRSgit — APPROVED** [review]
